File: 数组/优先堆&双指针.c
class Solution {
public:
    long long totalCost(vector<int>& costs, int k, int candidates) {
        int n = costs.size();
        if (n - 2 * candidates < k) {
            ranges::nth_element(costs, costs.begin() + k);  // 排序 ranges::nth_element(data, end)
            // 0LL 从0开始，为long long类型，相加，accumulate(start, end, type)
            return accumulate(costs.begin(), costs.begin() + k, 0LL);
        }

        priority_queue<int, vector<int>, greater<>> pre, suf;  // 优先堆
        for (int i = 0; i < candidates; i++) {
            pre.push(costs[i]);
            suf.push(costs[n - i - 1]);
        }

        // 双指针
        long long ans = 0;
        int i = candidates, j = n - candidates - 1;
        while (k--) {
            if (pre.top() <= suf.top()) {
                ans += pre.top();
                pre.pop();
                pre.push(costs[i++]);
            } else {
                ans += suf.top();
                suf.pop();
                suf.push(costs[j--]);
            }
        }
        return ans;
    }
};

File: 数组/优先堆&双指针.java
class Solution {
    public long totalCost(int[] costs, int k, int candidates) {
        int n = costs.length;
        long ans = 0;
        if (n - 2 * candidates < k) {
            Arrays.sort(costs);  // 数组排序
            for (int i = 0; i < k; i++) {
                ans += costs[i];
            }
            return ans;
        }

        PriorityQueue<Integer> pre = new PriorityQueue<>();  // 初始化优先队列 '()' 非常重要
        PriorityQueue<Integer> suf = new PriorityQueue<>();  //
        for (int i = 0; i < candidates; i++) {
            pre.offer(costs[i]);
            suf.offer(costs[n - i - 1]);
        }

        // 双指针
        int i = candidates;
        int j = n - candidates - 1;
        while (k-- > 0) {
            if (pre.peek() <= suf.peek()) {  // 获取第一个数据 .peek()
                ans += pre.poll();
                pre.offer(costs[i++]);
            } else {
                ans += suf.poll();
                suf.offer(costs[j--]);
            }
        }
        return ans;
    }
}